### ISSUE TITLE
Always prefer exception from `thing` action over exception from `after`

### ISSUE THREAD

**Issue by Yuras** [fpco/safe-exceptions#2]
You decided to run cleanup action under `uninterruptibleMask_`, so it can't throw async exception. So checking for async exception [here](https://github.com/fpco/safe-exceptions/blob/750195160511ebad99a85c41975cd9f24baaad93/src/Control/Exception/Safe.hs#L215) makes little sense. Just ignore exception from `after` if `thing` already failed, either with sync or async exception.

To be fair, I know one, and only one, way to throw async exception under `uninterruplibleMask_` -- `throwTo` the calling thread, see [docs](https://hackage.haskell.org/package/base-4.9.0.0/docs/Control-Concurrent.html#v:throwTo). But I'd not account this use case -- it will break all code anyway.

Probably not directly related: [here](http://blog.haskell-exists.com/yuras/posts/handling-async-exceptions-in-haskell-pushing-bracket-to-the-limits.html#example-5-uninterruptiblemask) is an example how interruptible cleanup action could be useful.

**Comment by mitchellwrosen** [comment]
I agree, I think the only way that check makes sense is if the thread throws an asynchronous exception synchronously. But then, there's probably no reason to prefer this exception over the first one thrown, and moreover I think the entire library depends upon users distinguishing between synchronous and asynchronous exceptions accurately.

```haskell
bracket :: C.MonadMask m => m a -> (a -> m b) -> (a -> m c) -> m c
bracket before after thing = C.mask $ \restore -> do
    x <- before
    res1 <- C.try $ restore (thing x)
    C.uninterruptibleMask_ $ after x
    case res1 of
        Left (e1 :: SomeException) -> C.throwM e1
        Right y -> return y
```

**Comment by snoyberg** [comment]
The concern I'd have with removing the check is that it's still technically possible for exceptions with an async type to be thrown synchronously (via `Control.Exception.throwIO`) from within the exception handler, and I'm not sure what we would want to do in that case. Removing the check we have now and just rethrowing `thing`'s exception may end up with different behavior than today.

**Comment by snoyberg** [comment]
@mitchellwrosen Hahaha, that timing was amazing.

**Comment by Yuras** [comment]
Removing the check lets you provide simple and clean description of how exceptions are handled in cleanup actions instead of vague guarantees you have right now. It is up to you to decide of course.

Throwing async exception using `throwIO` is as silly as not releasing resource in the cleanup action at all. You can't prevent the later, right? So just document how correct cleanup action should behave.

If you are using `uninterruptibleMask_`, then you assume the code will break in case of async exception, otherwise why do you need `uninterruptibleMask_` in the first place? If user throws async exception via `throwIO`, then your code is broken, and it is a bug in cleanup action.

Lets put it into wider context. If you accept the commonly used approach to [exception safety](https://en.wikipedia.org/wiki/Exception_safety), then the only real use case for `uninterruplibleMask_` is to get "no-throw" safety level. It means that you already eliminated all possible ways to fail via sync exception, and now want to eliminate async exception too. It is much rarer situation then people usually claim. You can do it for `readIORef` and `writeIORef`, `takeMVar` and `putMVar` (but only if you can guarantee they can't block indefinitely), simple `STM` actions that don't throw, and it is probably that. Using `uninterruplibleMask_` with `hClose`, `waitForProcess` or most of other `IO` actions makes no sense.

You also are using `uninterruptibleMask_` to resolve the double throw issue (what to do if both `thign` and `after` throw). It is probably not the only way to resolve it, but anyway it makes sense (it is better then `Control.Exception`, which is simply broken). So strictly speaking you don't need `uninterruptibleMask_` when `thing` doesn't fail. I'd recommend the next version of `bracket`:

```haskell
bracket :: C.MonadMask m => m a -> (a -> m b) -> (a -> m c) -> m c
bracket before after thing = C.mask $ \restore -> do
    x <- before
    res <- C.try $ restore (thing x)
    case res of
        Left (e :: SomeException) -> do
          C.uninterruptibleMask_ $ after x
          C.throwM e
        Right y -> do
          after x
          return y
```

**Comment by snoyberg** [comment]
I buy these arguments, I'm OK with simplifying and then having the much simpler semantics. I may still hold onto the vague wording in the README, though, since I can definitely see an argument for having rolled-up exceptions types (combining two different exceptions) in the future. But I don't want to tackle that right now unless someone wants to back me up on its usefulness.

Is it intentional that your second call to `after` is not `uninterruptibleMask_`d? That's a topic that I discussed with @peaker on Twitter. I have I feeling the two of you disagree on whether this is a good idea or not. Link to relevant Twitter discussion: https://twitter.com/EyalL/status/745516001165381632

**Comment by ElastiLotem** [comment]
Note that even if an IO action may throw sync exceptions, it may indicate the cleanup is either not needed (resource already vanished) or impossible.

Even if it is the latter, it means we cannot reach the nothrow level of safety, we may still prefer to leak a resource in the rare circumstance that releasing it becomes impossible while not leaking it when getting the much more frequent async exceptions.

In other words, not leaking upon async exceptions is beneficial whether or not you manage to avoid leaks in your sync exceptions.

**Comment by Yuras** [comment]
We agree with @Peaker in general -- exception handling in Haskell is broken. We [disagree in details](https://twitter.com/EyalL/status/720966203632697344), e.g. whether the second call to `after` should be surrounded with `unterruptibleMask_` or not. As I wrote, I think `uninterruptibleMask_` is rarely necessary, he argues the opposite.

**Comment by snoyberg** [comment]
Let's move the interruptible vs uninterruptible discussion to #3. As far as this issue goes, I think we're all in agreement that changing the behavior to always rethrow `thing`'s exception is the right thing to do, which I'll go ahead with.

**Comment by snoyberg** [comment]
Please see PR #5
